### PR TITLE
Refactor anomalib to new annotation format, add `refurb` and `pyupgrade`

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- Refactor anomalib to new annotation format
- Modify the code using `refurb`. The tool has not yet been added to the pre-commit configuration since it requires Python3.10 to execute.
- Add pyupgrade to pre-commit settings and modify the code accordingly.

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).

### CODE REVIEW

File: anomalib/data/avenue.py [hunk 1]
```diff
@@ -172,23 +173,23 @@ class Avenue(AnomalibVideoDataModule):
         clip_length_in_frames (int, optional): Number of video frames in each clip.
         frames_between_clips (int, optional): Number of frames between each consecutive video clip.
         task TaskType): Task type, 'classification', 'detection' or 'segmentation'
-        image_size (Optional[Union[int, Tuple[int, int]]], optional): Size of the input image.
+        image_size (int | tuple[int, int] | None, optional): Size of the input image.
```
Comment (djdameln) [inline anomalib/data/avenue.py:176]: Could you explain why we switch from `Tuple[int, int]` to `tuple[int, int]`?

Comment (samet-akcay) [reply anomalib/data/avenue.py:176]: With the new annotations, there is no need for `from typing import Tuple`. We could just do `tuple[int, int]`

File: anomalib/models/cfa/lightning_model.py [hunk 2]
```diff
@@ -73,43 +74,40 @@ def on_train_start(self) -> None:
         """Initialize the centroid for the memory bank computation."""
         self.model.initialize_centroid(data_loader=self.trainer.datamodule.train_dataloader())  # type: ignore
 
-    def training_step(self, batch) -> STEP_OUTPUT:
+    def training_step(self, batch: dict[str, str | Tensor], *args, **kwargs) -> STEP_OUTPUT:
```
Comment (djdameln) [inline anomalib/models/cfa/lightning_model.py:77]: Is it really needed to add `*args` and `**kwargs` here and in the other lightning models?

File: anomalib/models/components/base/anomaly_module.py [hunk 3]
```diff
@@ -32,14 +35,14 @@ class AnomalyModule(pl.LightningModule, ABC):
...
-    def __init__(self):
+    def __init__(self) -> None:
         super().__init__()
         logger.info("Initializing %s model.", self.__class__.__name__)
 
         self.save_hyperparameters()
         self.model: nn.Module
-        self.loss: Tensor
-        self.callbacks: List[Callback]
+        self.loss: nn.Module
```
Comment (djdameln) [inline anomalib/models/components/base/anomaly_module.py:44]: I don't understand this change. As far as I know, the `loss` attribute holds the value of the loss between epochs, so it should be a Tensor.

File: anomalib/models/components/base/anomaly_module.py [hunk 4]
```diff
@@ -50,78 +53,81 @@ def __init__(self):
...
+            batch (Any): Current batch
             batch_idx (int): Index of current batch
-            _dataloader_idx (int): Index of the current dataloader
+            dataloader_idx (int): Index of the current dataloader
 
         Return:
             Predicted output
         """
-        outputs = self.validation_step(batch, batch_idx)
+        del batch_idx, dataloader_idx  # These variables are not used.
```
Comment (djdameln) [inline anomalib/models/components/base/anomaly_module.py:85]: can't we solve this with `_batch_idx` and `_dataloader_idx`

File: anomalib/models/components/feature_extractors/timm.py [hunk 5]
```diff
@@ -22,7 +23,7 @@ class TimmFeatureExtractor(nn.Module):
 
     Args:
         backbone (nn.Module): The backbone to which the feature extraction hooks are attached.
-        layers (Iterable[str]): List of layer names of the backbone to which the hooks are attached.
+        layers (Iterable[str]): list of layer names of the backbone to which the hooks are attached.
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/timm.py:26]: Minor detail, but I think we can keep this capitalized, because it's the first word of the sentence.

File: setup.py [hunk 6]
```diff
@@ -48,27 +50,27 @@ def get_version() -> str:
...
-        List[str]: List of required packages
+        list[str]: List of required packages
     """
 
-    required_packages: List[str] = []
+    required_packages: list[str] = []
 
     for requirement_file in requirement_files:
-        with open(f"requirements/{requirement_file}.txt", "r", encoding="utf8") as file:
+        with open(f"requirements/{requirement_file}.txt", encoding="utf8") as file:
```
Comment (djdameln) [inline setup.py:73]: Why this change?

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks. Much more readable now. I have a few comments. My main concern is with the deletion of unused attributes with the `del` keyword, which feels a bit hacky. Maybe we could think of an alternative approach.

File: anomalib/models/cfa/lightning_model.py (hunk 2 above)
Comment (samet-akcay) [reply anomalib/models/cfa/lightning_model.py:77]: > Is it really needed to add `*args` and `**kwargs` here and in the other lightning models?

The original `LightningModule` has this signature. Looks like this is the proper way of inheritance to satisfy pylint. Otherwise pylint throws `arguements-differ` error.

File: anomalib/data/btech.py [hunk 7]
```diff
@@ -192,12 +193,12 @@ class BTech(AnomalibDataModule):
...
         transform_config_val: Config for pre-processing during validation.
         create_validation_set: Create a validation subset in addition to the train and test subsets
-        seed (Optional[int], optional): Seed used during random subset splitting.
+        seed (int | None, optional): Seed used during random subset splitting.
         test_split_mode (TestSplitMode): Setting that determines how the testing subset is obtained.
         test_split_ratio (float): Fraction of images from the train set that will be reserved for testing.
         val_split_mode (ValSplitMode): Setting that determines how the validation subset is obtained.
         val_split_ratio (float): Fraction of train or test images that will be reserved for validation.
-        seed (Optional[int], optional): Seed which may be set to a fixed value for reproducibility.
+        seed (int | None, optional): Seed which may be set to a fixed value for reproducibility.
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/data/btech.py:201]: > My main concern is with the deletion of unused attributes with the del keyword, which feels a bit hacky. 

This is based on Google's Python style guidelines.
https://google.github.io/styleguide/pyguide.html#214-decision

File: anomalib/models/components/base/anomaly_module.py (hunk 3 above)
Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:44]: in almost all of the implementations, we use `self.loss` to compute the loss. For example, here
https://github.com/openvinotoolkit/anomalib/blob/ce70716dbcb8fca2e2b2437eed356c0d0d9f6ca3/anomalib/models/draem/lightning_model.py#L41

When we define this as `Tensor`, pylint and mypy complains, because we use this as `Callable`

Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:44]: If PL uses this under the hood, we could then rename this to `self.loss_func`

File: anomalib/models/components/base/anomaly_module.py (hunk 4 above)
Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:85]: Again this is based on this style guide, but if we agree to use _batch_idx, then I'm fine. Of course we need to ensure if pylint is happy.
https://google.github.io/styleguide/pyguide.html#214-decision

File: anomalib/models/csflow/lightning_model.py [hunk 8]
```diff
@@ -90,21 +92,21 @@ class CsflowLightning(Csflow):
...
             cross_conv_hidden_channels=hparams.model.cross_conv_hidden_channels,
             clamp=hparams.model.clamp,
             num_channels=3,
         )
-        self.hparams: Union[DictConfig, ListConfig]  # type: ignore
+        self.hparams: DictConfig | ListConfig  # type: ignore
         self.save_hyperparameters(hparams)
 
-    def configure_callbacks(self):
+    def configure_callbacks(self) -> list[EarlyStopping]:
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/models/csflow/lightning_model.py:109]: Yes! Thanks for spotting!

File: anomalib/models/components/base/anomaly_module.py (hunk 4 above)
Comment (samet-akcay) [reply anomalib/models/components/base/anomaly_module.py:85]: The guideline states the following though
> Other common forms of suppressing this warning include using ‘_’ as the identifier for the unused argument or prefixing the argument name with ‘unused_’, or assigning them to ‘_’. These forms are allowed but no longer encouraged. These break callers that pass arguments by name and do not enforce that the arguments are actually unused.

File: anomalib/utils/cli/cli.py [hunk 9]
```diff
@@ -47,19 +49,19 @@ class AnomalibCLI(LightningCLI):
 
     def __init__(  # pylint: disable=too-many-function-args
         self,
-        model_class: Optional[Union[Type[LightningModule], Callable[..., LightningModule]]] = None,
-        datamodule_class: Optional[Union[Type[LightningDataModule], Callable[..., LightningDataModule]]] = None,
-        save_config_callback: Optional[Type[SaveConfigCallback]] = SaveConfigCallback,
+        model_class: type[LightningModule] | Callable[..., LightningModule] | None = None,
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/utils/cli/cli.py:52]: That's PL code. I guess that's their preference. Not sure about the actual intention

File: setup.py (hunk 6 above)
Comment (samet-akcay) [reply setup.py:73]: `refurb` suggests this change.

`[FURB101]: Use `y = Path(x).read_text()` instead of `with open(x, ...) as f: y = f.read()``

File: anomalib/data/btech.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/data/btech.py:201]: Alright, in that case I'm fine with it. It still feels a bit hacky but maybe I just need to get used to it :)

[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/data/btech.py:201]: Yeah, initially looked a bit weird to me too :D

File: anomalib/utils/cli/cli.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/utils/cli/cli.py:52]: When I removed `type` I cannot silence some mypy issues. I guess we might need to explicitly specify it here.

File: setup.py (hunk 6 above)
Comment (samet-akcay) [reply setup.py:73]: This is an automatic change by `pyupgrade`, meaning that we don't need `"r"` in recent versions of python.

File: anomalib/utils/cli/cli.py [hunk 10]
```diff
@@ -147,7 +149,7 @@ def __set_default_root_dir(self) -> None:
         # If `resume_from_checkpoint` is not specified, it means that the project has not been created before.
         # Therefore, we need to create the project directory first.
         if config.trainer.resume_from_checkpoint is None:
-            root_dir = config.trainer.default_root_dir if config.trainer.default_root_dir else "./results"
+            root_dir = config.trainer.default_root_dir or "./results"
```
Comment (djdameln) [inline anomalib/utils/cli/cli.py:152]: cool, I was not aware of this syntax

**Review by djdameln — APPROVED** [review]
Thanks, just one minor comment, which could probably be ignored

File: anomalib/utils/cli/cli.py (hunk 10 above)
Comment (samet-akcay) [reply anomalib/utils/cli/cli.py:152]: me neither :)

File: anomalib/models/csflow/torch_model.py [hunk 11]
```diff
@@ -400,50 +400,59 @@ def __init__(
...
-                Node(
-                    [nodes[-1].out0, nodes[-1].out1, nodes[-1].out2],
-                    ParallelGlowCouplingLayer,
-                    {
-                        "clamp": self.clamp,
-                        "subnet_args": {
-                            "channels_hidden": self.cross_conv_hidden_channels,
-                            "kernel_size": self.kernel_sizes[coupling_block],
-                        },
+            nodes.extend([permute_node])
```
[the remark this answers is not in the document]
Comment (samet-akcay) [reply anomalib/models/csflow/torch_model.py:425]: I feel so too, but just did it this way to silence refurb :)
